### PR TITLE
Core: Pass denormalized stories to the sort function

### PR DESCRIPTION
**Pull request by tmeasday**
Issue: #11010 

## What I did

Denormalize stories before sorting

## How to test

- Is this testable with Jest or Chromatic screenshots? YEP

### TODO

I think I should memoize the denorming part otherwise there is no point in memoizing the sorting.

### CODE REVIEW

File: lib/client-api/src/story_store.ts [hunk 1]
```diff
@@ -503,33 +503,37 @@ export default class StoryStore {
       .map((i) => this.mergeAdditionalDataToStory(i));
   }
 
-  sortedStories(): StoreItem[] {
+  sortedStories(options: { normalizeParameters?: boolean } = {}): StoreItem[] {
+    // We need to pass the stories with denormalized parameters to the sort function (see #11010)
+    const denormalizedStories = mapValues(this._stories, (story) => ({
```
Comment (shilman) [inline lib/client-api/src/story_store.ts:508]: Don't do the work if normalizeParameters is passed?

```js
    const stories = normalizeParameters ? this._stories : mapValues(this._stories, (story) => ({ ...
```

File: lib/client-api/src/story_store.ts [hunk 2]
```diff
@@ -503,33 +503,37 @@ export default class StoryStore {
...
-    return getSortedStories(this._stories, kindOrder, storySortParameter);
+    const orderedIds = getSortedStoryIds(
+      denormalizedStories,
+      kindOrder,
+      storySortParameter,
+      options
+    );
+
+    const storiesToReturn = options.normalizeParameters ? this._stories : denormalizedStories;
+    return orderedIds.map((id) => storiesToReturn[id]);
```
Comment (shilman) [inline lib/client-api/src/story_store.ts:525]: `stories[id]` with the change above ☝️

**Review by shilman — CHANGES_REQUESTED** [review]
Minor suggestions.

I didn't get the memoization part. Can you elaborate?

File: lib/client-api/src/story_store.ts (hunk 1 above)
Comment (tmeasday) [reply lib/client-api/src/story_store.ts:508]: We need to pass the denormalised stories to the sort function no matter what

Comment (shilman) [reply lib/client-api/src/story_store.ts:508]: 💡

**Comment by tmeasday** [comment]
`sortedStories()` calls `getSortedStoryIds` which is memoized (once). But the first argument is an object that is recreated each time you call `sortedStories`. So I think that means it'll never reuse the memoized result, which makes the whole thing pointless. I'll double check.

**Comment by tmeasday** [comment]
ok the memoizing wasn't doing anything to start off with.

**Review by shilman — APPROVED** [review]
LGTM!

**Pull request merged**